x/validators/handler.go

```diff
@@ -81,32 +79,28 @@ func (h updateHandler) validate(ctx weave.Context, store weave.KVStore, tx weave
 		return nil, resUpdates, errors.Wrap(errors.ErrUnauthorized, "no permission")
 	}
 
-	updates, err := GetValidatorUpdates(store)
+	updates, err := weave.GetValidatorUpdates(store)
 	if err != nil {
 		return nil, resUpdates, errors.Wrap(err, "failed to query validators")
 	}
 
 	resUpdates = updates
-	validatorSlice := resUpdates.ValidatorUpdates
 
-DiffLoop:
 	for _, v := range diff {
-		for key, validator := range validatorSlice {
-			if bytes.Equal(v.PubKey.Data, validator.PubKey.Data) {
-				if v.Power == validator.Power {
-					return nil, resUpdates, errors.Wrap(errors.ErrInput, "same validator power")
-				}
-
-				resUpdates.ValidatorUpdates[key] = v
-				continue DiffLoop
+		if validator, key, ok := resUpdates.Get(v.PubKey); ok {
+			if v.Power == validator.Power {
+				return nil, resUpdates, errors.Wrap(errors.ErrInput, "same validator power")
 			}
-
+			resUpdates.ValidatorUpdates[key] = v
+			continue
 		}
+
 		if v.Power == 0 {
 			return nil, resUpdates, errors.Wrap(errors.ErrInput, "setting unknown validator power to 0")
 		}
+
 		resUpdates.ValidatorUpdates = append(resUpdates.ValidatorUpdates, v)
 	}
 
-	return diff, resUpdates, nil
+	return diff, resUpdates.Deduplicate(true), nil
```

Comment:
pp: this is not the same as preventing duplicates in the incoming data. IMHO we should check that and fail but I was not involved in previous discussions. My main reason is for consistency with other handlers where we do this and simple api doc.

Reply:
I don't think you understand what this does, which is fair, as this is an opinionated approach. This basically prepares the set for storage, it used to be in store, but now it's called explicitly everywhere. This data is used for bookkeeping only.

Reply:
Please don't make assumptions what you think I understood or what not. The point I raised was that you deduplicate the result rather than preventing duplicates in the incoming message.